### ISSUE TITLE
External flash encrypted with Chacha broken

### ISSUE THREAD

**Issue by LOorts-Aloxy** [wolfSSL/wolfBoot#211]
Hi, we are updating our bootloader to use the newest version but it seems the encrypted bootloader is broken for us starting from commit [b7026a5b1cbd7428eba3929a101ff198624e26c8](https://github.com/wolfSSL/wolfBoot/commit/b7026a5b1cbd7428eba3929a101ff198624e26c8). Right before this commit, everything seems to work fine. But on this commit (and any from this point on), our update fails.

**Comment by LOorts-Aloxy** [comment]
Hi @rizlik ,
We don't use a custom user_settings.h.
The config settings we use are the following:
`
  ARCH=ARM
  TARGET=stm32l0
  SIGN=ED25519
  HASH=SHA256
  DEBUG=0
  OPTIMIZATION_LEVEL=1
  VTOR=1
  NO_ASM=0
  SPI_FLASH=1
  ENCRYPT=1
  NO_MPU=1
  ALLOW_DOWNGRADE=0
  NVM_FLASH_WRITEONCE=1
  WOLFBOOT_VERSION=0x00000400
  V=0
  SPMATH=1
  RAM_CODE=1
  DUALBANK_SWAP=0
  IMAGE_HEADER_SIZE=256
  WOLFBOOT_PARTITION_SIZE=0x2A000
  WOLFBOOT_SECTOR_SIZE=0x1000
  WOLFBOOT_PARTITION_BOOT_ADDRESS=0x5000
  WOLFBOOT_PARTITION_UPDATE_ADDRESS=0x35000
  WOLFBOOT_PARTITION_SWAP_ADDRESS=0x3000
  WOLFBOOT_LOAD_ADDRESS=0x200000 
  WOLFBOOT_LOAD_DTS_ADDRESS=0x400000
`

**Comment by LOorts-Aloxy** [comment]
Hi @rizlik, 
I can confirm that the key tools are also updated when updating to newer commits. Commit https://github.com/wolfSSL/wolfBoot/commit/61275ec9dde2a72dbdf4085b8e03ce301d91a703 still fails to update the device using those settings. FYI, we also have a UART, not-encrypted bootloader that does work on both commits.

**Comment by rizlik** [comment]
Hi @LOorts-Aloxy,
unfortunately, we couldn't pinpoint the problem. Our automated tests on real hardware and a new test on a simulated target don't show any issue. If you can provide us with further debugging information that will greatly help (especially when the update process fails and why). I'll try to get an stm32l0 hardware by next week and reproduce the bug there.

**Comment by LOorts-Aloxy** [comment]
Did you also test (in simulation and/or on real hardware) with encryption on an external SPI flash? I don't see any tests/examples in the repo.

**Comment by rizlik** [comment]
we have tested using an external flash, but not using an SPI layer. The SPI layer wasn't modified lately, so the problem shouldn't be there. You can see the tests regarding the encryption here: https://github.com/wolfSSL/wolfBoot/blob/master/tools/test-enc.mk

**Comment by danielinux** [comment]
Hi @LOorts-Aloxy - we are busy producing more tests. Can you tell the size of your update image `image_v5_signed_and_encrypted.bin` ?

Thanks,

-- 
Daniele

**Comment by LOorts-Aloxy** [comment]
Hi @danielinux,
Thank you for still proceeding with this. 

The size of ```image_v5_signed_and_encrypted.bin``` is 0x23830.

**Comment by LOorts-Aloxy** [comment]
Hi all. I've had the chance to test this a bit further. 


side notes: our image is on an spi external flash. The address of the update partition is 0x35000 and the size of the image is 0x2A000.

**Working**
On commit https://github.com/wolfSSL/wolfBoot/commit/644d83309e899b784da9a46b411c12dc6bf272cb, the update is succeeding:
<img width="1071" alt="image" src="https://user-images.githubusercontent.com/50589042/176698900-21854bd4-309e-44a4-8ffb-c2f4c5f61fd8.png">
(0x03 is a read command followed by a 3 bytes address 0x05EFD0 and 0x05EFCE, the other party then answers with bytes)
the screenshot shows the magic number getting read out and the partition state. This indicates IMG_STATE_UPDATING so it starts the update.

**Not working**
On commit https://github.com/wolfSSL/wolfBoot/commit/67912790de218f949609e6cdd8e04654dd1b891f, the update is failing:
![MicrosoftTeams-image](https://user-images.githubusercontent.com/50589042/176699992-9417fd5d-881c-44c6-a2dd-4a7e37d8c9c5.png)
Both show the exact same interaction but the working commit then proceeds to read the first sector again while the not-working commit has no SPI interactions anymore until it boots again on its original firmware.



Steps we see after each other on SPI:
- read at 0x5EFD0 of 4 bytes (answer 0x42, 0x4F, 0x4F, 0x54)
- read at 0x5EFCF of 4 bytes (answer 0x70, 0x42, 0x4F, 0x4F)
- Read at 0x35000 of 0x100 bytes
- another read at 0x35000 of 0x100 bytes
- read at 0x5EFD0 of 4 bytes (answer 0x42, 0x4F, 0x4F, 0x54)
- read at 0x5EFCF of 4 bytes (answer 0x70, 0x42, 0x4F, 0x4F)
- read at 0x35000 of 0x100 bytes
- another read at 0x35000 of 0x100 bytes
- read at 0x5EFD0 of 4 bytes (answer 0x42, 0x4F, 0x4F, 0x54)
- read at 0x5EFCE of 4 bytes (answer 0xFF, 0x70, 0x42, 0x47)
working:
- read at 0x35000 of 0x100 bytes
not working:
- Then there is no SPI communication for 11 seconds and our own firmware starts again

Hope this helps to reproduce/debug this. I updated on both commits with exactly the same encrypted binary file so it seems it can't be anything with the signing tool. The data itself also doesn't show anything weird, one commit just stops with all interactions, the other proceeds as normal.

**Comment by danielinux** [comment]
@LOorts-Aloxy from the symptoms it still looks like the decrypt of the first sector is failing in the second case.

Are you 100% positive that the sign.py tool that create the second image is updated to (at least ) 61275ec9dde2a72dbdf4085b8e03ce301d91a703 ? You mentioned you are using the same binary, but there is a difference in how the IV is incremented by `sign.py` when creating the new image starting from that commit. In other words, due to a bug in the IV calculation, fixed in b7026a5, we had to break the compatibility with images created with a `sign.py` older than 61275ec9dde and `sign` (C version) older than `

I can reproduce the same issue with a very similar behavior by using an image created with an old `sign.py` tool and a newer wolfboot.

Could you try and reproduce with a fresh clone from master? Or perhaps using the equivalent `sign` C tool instead of `sign.py` (compiled using `make -C tools/keytools`).

Please note that C keytools have been fixed to use the new IV counters only from 67912790 on.

**Comment by danielinux** [comment]
My flash geometry:
```
WOLFBOOT_PARTITION_SIZE?=0x20000
WOLFBOOT_SECTOR_SIZE?=0x1000
WOLFBOOT_PARTITION_BOOT_ADDRESS?=0x08010000
WOLFBOOT_PARTITION_UPDATE_ADDRESS?=0x0
WOLFBOOT_PARTITION_SWAP_ADDRESS?=0x20000
```

Working update from latest master (verbose, logging all read/write/erase):

```
Version running on target: 1
[\] Fetching update blocks 			Read @0
Boot partition version from test app: 1
[|] Fetching update blocks 			Read @1ffd0
[/] Writing backup blocks  			Write @1ffd0
[-] Fetching update blocks 			Read @1ffcf
[\] Writing backup blocks  			Write @1ffcf
st-flash 1.7.0-11-ge662da0-dirty
2022-07-01T08:38:59 INFO common.c: WB55: 256 KiB SRAM, 1024 KiB flash in at least 4 KiB pages.

** TARGET REBOOT **
Version running on target: 1
[|] Fetching update blocks 			Read @1ffd0
[/] Fetching update blocks 			Read @1ffcf
[-] Fetching update blocks 			Read @0
[\] Fetching update blocks 			Read @0
[|] Fetching update blocks 			Read @1ffd0
[/] Fetching update blocks 			Read @1ffcf
[-] Fetching update blocks 			Read @0
[\] Fetching update blocks 			Read @0
[|] Fetching update blocks 			Read @1ffd0
[/] Fetching update blocks 			Read @1ffce
[-] Fetching update blocks 			Read @0
[\] Fetching update blocks 			Read @100
[|] Fetching update blocks 			Read @200
[/] Fetching update blocks 			Read @300
[-] Fetching update blocks 			Read @400
[\] Fetching update blocks 			Read @500
[|] Fetching update blocks 			Read @600
[/] Fetching update blocks 			Read @700
[-] Fetching update blocks 			Read @800
[\] Fetching update blocks 			Read @900
[|] Fetching update blocks 			Read @a00
[/] Fetching update blocks 			Read @b00
[-] Fetching update blocks 			Read @c00
[\] Fetching update blocks 			Read @d00
[|] Fetching update blocks 			Read @e00
[/] Fetching update blocks 			Read @f00
[-] Fetching update blocks 			Read @1000
[\] Fetching update blocks 			Read @1100
[|] Fetching update blocks 			Read @1200
[/] Fetching update blocks 			Read @1300
[-] Fetching update blocks 			Read @0
[\] Fetching update blocks 			Read @1ffd0
[|] Fetching update blocks 			Read @1ffce
[/] Erase swap blocks      			Erase @20000
[-] Fetching update blocks 			Read @0
[\] Writing SWAP blocks    			Write @20000
[|] Fetching update blocks 			Read @1ffd0
[/] Fetching update blocks 			Read @1ffce
[-] Writing backup blocks  			Write @1ffce
[\] Erase update blocks    			Erase @0
[|] Writing backup blocks  			Write @0
[/] Fetching update blocks 			Read @1ffd0
[-] Fetching update blocks 			Read @1ffce
[\] Writing backup blocks  			Write @1ffce
[|] Reading SWAP blocks    			Read @20000
[/] Fetching update blocks 			Read @1ffd0
[-] Fetching update blocks 			Read @1ffce
[\] Writing backup blocks  			Write @1ffce
[|] Fetching update blocks 			Read @1ffd0
[/] Fetching update blocks 			Read @1ffce
[-] Erase swap blocks      			Erase @20000
[\] Fetching update blocks 			Read @1000
[|] Writing SWAP blocks    			Write @20000
[/] Fetching update blocks 			Read @1ffd0
[-] Fetching update blocks 			Read @1ffce
[\] Writing backup blocks  			Write @1ffce
[|] Erase update blocks    			Erase @1000
[/] Writing backup blocks  			Write @1000
[-] Fetching update blocks 			Read @1ffd0
[\] Fetching update blocks 			Read @1ffce
[|] Writing backup blocks  			Write @1ffce
[/] Reading SWAP blocks    			Read @20000
[-] Fetching update blocks 			Read @1ffd0
[\] Fetching update blocks 			Read @1ffce
[|] Writing backup blocks  			Write @1ffce
[/] Erase update blocks    			Erase @2000
[-] Erase update blocks    			Erase @3000
[\] Erase update blocks    			Erase @4000
[|] Erase update blocks    			Erase @5000
[/] Erase update blocks    			Erase @6000
[-] Erase update blocks    			Erase @7000
[\] Erase update blocks    			Erase @8000
[|] Erase update blocks    			Erase @9000
[/] Erase update blocks    			Erase @a000
[-] Erase update blocks    			Erase @b000
[\] Erase update blocks    			Erase @c000
[|] Erase update blocks    			Erase @d000
[/] Erase update blocks    			Erase @e000
[-] Erase update blocks    			Erase @f000
[\] Erase update blocks    			Erase @10000
[|] Erase update blocks    			Erase @11000
[/] Erase update blocks    			Erase @12000
[-] Erase update blocks    			Erase @13000
[\] Erase update blocks    			Erase @14000
[|] Erase update blocks    			Erase @15000
[/] Erase update blocks    			Erase @16000
[-] Erase update blocks    			Erase @17000
[\] Erase update blocks    			Erase @18000
[|] Erase update blocks    			Erase @19000
[/] Erase update blocks    			Erase @1a000
[-] Erase update blocks    			Erase @1b000
[\] Erase update blocks    			Erase @1c000
[|] Erase update blocks    			Erase @1d000
[/] Erase update blocks    			Erase @1e000
[-] Erase update blocks    			Erase @1f000
[\] Erase swap blocks      			Erase @20000
[|] Fetching update blocks 			Read @0
Boot partition version from test app: 2
st-flash 1.7.0-11-ge662da0-dirty
2022-07-01T08:39:59 INFO common.c: WB55: 256 KiB SRAM, 1024 KiB flash in at least 4 KiB pages.

** TARGET REBOOT **
Version running on target: 2
[/] Fetching update blocks 			Read @0
Boot partition version from test app: 2
st-flash 1.7.0-11-ge662da0-dirty
2022-07-01T08:40:02 INFO common.c: WB55: 256 KiB SRAM, 1024 KiB flash in at least 4 KiB pages.
2022-07-01T08:40:02 INFO common.c: read from address 0x08010000 size 32768
5056+0 records in
5056+0 records out
5056 bytes (5.1 kB, 4.9 KiB) copied, 0.0904494 s, 55.9 kB/s
TEST SUCCESSFUL
```

same test `make test-enc-update`, running with wolfBoot from latest master, older `sign.py` (from b7026a5):


```
Version running on target: 1
[\] Fetching update blocks 			Read @0
Boot partition version from test app: 1
[|] Fetching update blocks 			Read @1ffd0
[/] Writing backup blocks  			Write @1ffd0
Unrecognized command: 57
st-flash 1.7.0-11-ge662da0-dirty
2022-07-01T08:35:41 INFO common.c: WB55: 256 KiB SRAM, 1024 KiB flash in at least 4 KiB pages.

** TARGET REBOOT **
Version running on target: 1
[-] Fetching update blocks 			Read @1ffd0
[\] Fetching update blocks 			Read @1ffcf
[|] Fetching update blocks 			Read @1ffd0
[/] Fetching update blocks 			Read @1ffcf
[-] Fetching update blocks 			Read @0
Boot partition version from test app: 1
[\] Fetching update blocks 			Read @1ffd0
[|] Fetching update blocks 			Read @1ffcf
[/] Writing backup blocks  			Write @1ffcf
TIMEOUT    
st-flash 1.7.0-11-ge662da0-dirty
2022-07-01T08:36:41 INFO common.c: WB55: 256 KiB SRAM, 1024 KiB flash in at least 4 KiB pages.

** TARGET REBOOT **
Version running on target: 1
[-] Fetching update blocks 			Read @1ffd0
[\] Fetching update blocks 			Read @1ffcf
[|] Fetching update blocks 			Read @0
[/] Fetching update blocks 			Read @0
[-] Fetching update blocks 			Read @1ffd0
[\] Fetching update blocks 			Read @1ffcf
[|] Fetching update blocks 			Read @0
[/] Fetching update blocks 			Read @0
[-] Fetching update blocks 			Read @1ffd0
[\] Fetching update blocks 			Read @1ffce
[|] Fetching update blocks 			Read @0
Boot partition version from test app: 1
[/] Fetching update blocks 			Read @1ffd0
[-] Fetching update blocks 			Read @1ffcf
st-flash 1.7.0-11-ge662da0-dirty
2022-07-01T08:36:44 INFO common.c: WB55: 256 KiB SRAM, 1024 KiB flash in at least 4 KiB pages.
2022-07-01T08:36:44 INFO common.c: read from address 0x08010000 size 32768
5056+0 records in
5056+0 records out
5056 bytes (5.1 kB, 4.9 KiB) copied, 0.0912318 s, 55.4 kB/s
Binary files boot.bin and test-app/image_v2_signed.bin differ
TEST FAILED
```

**Comment by LOorts-Aloxy** [comment]
hi @danielinux @rizlik it seems our automatic testing system is messing up the signing progress. On the current master, I can update successfully on local tests so it seems this problem is purely caused by our own progress somewhere. Thank you for looking into this together and the help!

**Comment by danielinux** [comment]
@LOorts-Aloxy you're welcome! Thanks for double checking that everything works from our side. Feel free to contact us again with any questions.

**Comment by LOorts-Aloxy** [comment]
To come back on this issue, our automated CI seems to be generating images incorrectly with the sign.py tool. When I generate the signed_and_encrypted package locally it works, but when taken from the CI, it does not.
The main differences are:
| Automated CI  | local  |
| -------------- | -----  |
| python3.8.10       | python3.9.12 |
| wolfSSL 4.3.0 | wolfSSL5.3.0 |
| Ubuntu 20.04 | MacOS |

We use wolfssl directly to build wolfcrypt (using ```USE_LOCAL_WOLFSSL=/... python3 -m pip install wolfcrypt-py/.```)

I don't see any direct other differences.
Any idea why it's not working or what could be breaking it? I locally tried the older python version and it still worked so that difference seems negligible.

**Comment by LOorts-Aloxy** [comment]
Just tested with Debian instead of Ubuntu. This has a package of wolfSSL available of version 4.6.0. It seems a build of wolfcrypt using this version of wolfssl does generate a valid image.

**Comment by LOorts-Aloxy** [comment]
now updated our build ci to just use the newest version of wolfcrypt that is available on pip and it seems to work as well. 

So, conclusion:
It seems version 4.1.0 of wolfcrypt-py does not work when built with wolfssl version 4.3.0. Using a newer version (4.6.0 and higher) seems to fix this and using a pre-built version (5.3.0) also fixes it.

**Comment by L-I-Am** [comment]
Quick question concerning this breaking change in the sign.py tool, would it make sense to add an argument that signs with the old IV? We already have devices in the field that use this old version so any update we now release will need to be signed with both.
